src/hashes.rs

```diff
@@ -1,110 +1,305 @@
-/// List of types currently supported in the multihash spec.
-///
-/// Not all hash types are supported by this library.
-#[derive(PartialEq, Eq, Clone, Debug, Copy, Hash)]
-pub enum Hash {
+use blake2b_simd::Params as Blake2b;
+use blake2s_simd::Params as Blake2s;
+use digest::Digest;
+use sha1::Sha1 as Sha1Hasher;
+use sha2::{Sha256, Sha512};
+use tiny_keccak::{Hasher, Keccak, Sha3};
+
+use crate::digests::{wrap, Multihash, MultihashDigest};
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Code {
+    /// Make it possible to use a custom code that is not part of the enum yet
+    Custom(u64),
+    /// Hashes that are immplemented in this crate
+    Implemented(ImplementedCode),
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum ImplementedCode {
     /// Identity (Raw binary )
-    Identity,
+    Identity = 0x00,
     /// SHA-1 (20-byte hash size)
-    SHA1,
+    Sha1 = 0x11,
     /// SHA-256 (32-byte hash size)
-    SHA2256,
+    Sha2_256 = 0x12,
     /// SHA-512 (64-byte hash size)
-    SHA2512,
-    /// SHA3-512 (64-byte hash size)
-    SHA3512,
-    /// SHA3-384 (48-byte hash size)
-    SHA3384,
-    /// SHA3-256 (32-byte hash size)
-    SHA3256,
+    Sha2_512 = 0x13,
     /// SHA3-224 (28-byte hash size)
-    SHA3224,
+    Sha3_224 = 0x17,
+    /// SHA3-256 (32-byte hash size)
+    Sha3_256 = 0x16,
+    /// SHA3-384 (48-byte hash size)
+    Sha3_384 = 0x15,
+    /// SHA3-512 (64-byte hash size)
+    Sha3_512 = 0x14,
     /// Keccak-224 (28-byte hash size)
-    Keccak224,
+    Keccak224 = 0x1a,
     /// Keccak-256 (32-byte hash size)
-    Keccak256,
+    Keccak256 = 0x1b,
     /// Keccak-384 (48-byte hash size)
-    Keccak384,
+    Keccak384 = 0x1c,
     /// Keccak-512 (64-byte hash size)
-    Keccak512,
+    Keccak512 = 0x1d,
     /// BLAKE2b-256 (32-byte hash size)
-    Blake2b256,
+    Blake2b256 = 0xb220,
     /// BLAKE2b-512 (64-byte hash size)
-    Blake2b512,
+    Blake2b512 = 0xb240,
     /// BLAKE2s-128 (16-byte hash size)
-    Blake2s128,
+    Blake2s128 = 0xb250,
     /// BLAKE2s-256 (32-byte hash size)
-    Blake2s256,
-}
-
-impl Hash {
-    /// Get the corresponding hash code.
-    pub fn code(self) -> u16 {
-        match self {
-            Hash::Identity => 0x00,
-            Hash::SHA1 => 0x11,
-            Hash::SHA2256 => 0x12,
-            Hash::SHA2512 => 0x13,
-            Hash::SHA3224 => 0x17,
-            Hash::SHA3256 => 0x16,
-            Hash::SHA3384 => 0x15,
-            Hash::SHA3512 => 0x14,
-            Hash::Keccak224 => 0x1A,
-            Hash::Keccak256 => 0x1B,
-            Hash::Keccak384 => 0x1C,
-            Hash::Keccak512 => 0x1D,
-            Hash::Blake2b256 => 0xB220,
-            Hash::Blake2b512 => 0xB240,
-            Hash::Blake2s128 => 0xB250,
-            Hash::Blake2s256 => 0xB260,
+    Blake2s256 = 0xb260,
+}
```

Comment:
There is no way to go from these codes to the structs used for generating digests? This also limits functionality in some dynamic cases where if you get the code from calling `algorithm(&self) -> Code` you couldn't generate another hash from this. Just a thought, could have a function which takes the code as input and returns a generic type bounded on `DynMultihashDigest` to solve this problem?

Reply:
That's a great idea! See the latest commits for an implementation.